<!DOCTYPE html>
<html>
  <head>
    <title></title>
    <meta charset="utf-8">
    <link href="AMLclassifiers_files/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="styling/sydney-fonts.css" type="text/css" />
    <link rel="stylesheet" href="styling/sydney.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">


class: title-slide
background-image: url("styling/USydLogo-black.svg"), url("styling/title-image1.jpg")
background-position: 10% 90%, 100% 50%
background-size: 160px, 100% 100%

.content-box-purple[
# .black[Activity 1]
## Identifying Biomarkers and Making Predictions of Treatment Resistance
### Ellis Patrick
### 29 June 2018
]

---

## Activity Overview

- RNA-seq Data Set: Acute Myeloid Leukaemia Treatment Resistance

--

- ClassifyR: Biomarker Selection Methods and Cross-validation

--

- Differential Means and Differential Distribution Classifiers for Classifying AML Patients

---

## RNA-seq Data Set: Acute Myeloid Leukaemia (AML) Treatment Resistance

- Primary therapy resistance is a major problem in acute myeloid leukemia treatment. Approximately 20â€“30% of younger adult patients with AML and as many as 50% of older adults are refractory to induction treatment.

--

- Research findings are &lt;a href="http://www.haematologica.org/content/103/3/456" target="_blank"&gt;published&lt;/a&gt; in *Haematologica* in 2018.

--

- The data is available from &lt;a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE106291" target="_blank"&gt;GEO Browser&lt;/a&gt; as 250 `txt.gz` files or a Microsoft Excel file where the gene expression values were standardised to have a mean of 0 and variance of 1.

--



Import the prepared tab-separated text files of RNA-seq read counts and clinical data.


```r
readCounts &lt;- read.delim("data/counts.txt", check.names = FALSE)
readCounts &lt;- as.matrix(readCounts)
sampleInfo &lt;- read.delim("data/samples.txt", check.names = FALSE)
```

---

## Clinical Data

The cinical data provides information about seven different characteristics of the patients.


```r
# Add scroll box.
head(sampleInfo)
```

```
##       ID Gender Age  Response Survivial Time Status RUNX1-RUNX1T1 Fusion
## 1 GEO-13   Male  47 Resistant             32   Dead                   No
## 2 GEO-14   Male  38 Resistant            125   Dead                   No
## 3 GEO-15 Female  38 Resistant              1   Dead                   No
## 4 GEO-16   Male  37 Resistant             18   Dead                   No
## 5 GEO-17   Male  33 Resistant            143   Dead                   No
## 6 GEO-18   Male   7 Resistant            127  Alive                   No
##   RUNX1 Mutation
## 1            Yes
## 2            Yes
## 3            Yes
## 4            Yes
## 5             No
## 6            Yes
```

```r
nrow(sampleInfo)
```

```
## [1] 250
```

Observe the number of samples which are resistant to treatment and which are not.


```r
table(sampleInfo[, "Response"])
```

```
## 
## Resistant Sensitive 
##        71       164
```

`\(164 + 71 = 235\)`, so there are 15 patients with missing data regarding their resistance. Identify which rows of the clinical data they are in.


```r
removeClinical &lt;- which(is.na(sampleInfo[, "Response"]))
```

---

## AML Read Counts


```r
dim(readCounts)
```

```
## [1] 23367   250
```

There are 23367 genes in the counts table and 250 AML patients.

--


```r
readCounts[1:6, 1:6]
```

```
##          GEO-13 GEO-14 GEO-15 GEO-16 GEO-17 GEO-18
## A1BG        270     84     22    245    380    581
## A1BG-AS1     32     59      7     60    126     88
## A1CF          0      0      0      0      0      0
## A2LD1        14      4     18     19     35      1
## A2M          38      8     58    185      0     20
## A2ML1         0      0      1      0      0      0
```

---

## AML Read Counts

The number of reads per sample varies widely betwen patients by as much as 10-fold.


```r
samplesCounts &lt;- colSums(readCounts)
countsSummary &lt;- summary(samplesCounts)
countsSummary
```

```
##     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
##  3427284  9198530 10865612 12365192 13335215 41850915
```

--

Identify in which columns of the matrix patients with less than the first quartile of counted reads and more than the third quartile are to remove them from the analysis.


```r
removeRNA &lt;- which(samplesCounts &lt; countsSummary["1st Qu."] |
                   samplesCounts &gt; countsSummary["3rd Qu."])
```

---

## Subsetting the Data Set

Remove samples which have a number of read counts below the first quartile of counted reads or above the third quartile as well as those missing resistance status.


```r
allRemove &lt;- union(removeClinical, removeRNA)
readCounts &lt;- readCounts[, -allRemove]
sampleInfo &lt;- sampleInfo[-allRemove, ]
```

--

Caclulate the number of remanining samples belonging to each class.


```r
classes &lt;- sampleInfo[, "Response"]
table(classes)
```

```
## classes
## Resistant Sensitive 
##        40        73
```

--

There is a moderate amount of class imbalance in this data set.

---

## Relationship of Mean and Variance

- RNA-seq count measurements are known to have a relationship between mean and variance.

- Many classification algorithms require data with constant variance over the range of means.

Observe the mean-variance relationship after scaling counts for different total counts per sample.


```r
library(EDASeq)
AMLExpressionSet &lt;- newSeqExpressionSet(readCounts)
AMLExpressionSet &lt;- betweenLaneNormalization(AMLExpressionSet)
#meanVarPlot(AMLExpressionSet, log = TRUE, ylim = c(0, 20),
 #           main = "Mean-Variance Plot") bug; doesn't work on normalised counts!
```

---

Using DESeq2's `varianceStabilizingTransformation` function, remove the relationship of variance to mean.


```r
library(DESeq2)
measurementsVS &lt;- varianceStabilizingTransformation(readCounts)
```

--

Visualise again the relationship between mean and variance.


```r
meanVarPlot(newSeqExpressionSet(measurementsVS), log = TRUE,
            xlim = c(1, 3), ylim = c(0, 3), main = "Mean-Variance Plot")
```

```
## Warning in validityMethod(object): 'counts' contains non-integer numbers
```

![](AMLclassifiers_files/figure-html/unnamed-chunk-14-1.png)&lt;!-- --&gt;

The variance is now constant across the range of means.

## Subsetting the Data Set Again

Before doing any classification, the number of genes will be reduced to make feature selection faster. The top 2000 most variable genes are retained.


```r
geneVariances &lt;- apply(measurementsVS, 1, var)
mostVariable &lt;- order(geneVariances, decreasing = TRUE)[1:2000]
measurementsVS &lt;- measurementsVS[mostVariable, ]
measurementsVS[1:6, 1:6]
```

```
##             GEO-13    GEO-14    GEO-15    GEO-16    GEO-19    GEO-23
## XIST      5.062396  5.178148 13.541422  9.188697  9.618966 11.462734
## RPS4Y1   12.325286 11.771083  4.549248 12.194969 10.613627 11.691603
## HLA-DQB1  4.904264  6.236311 12.780571  7.973863 11.617717  8.386216
## HBG2      5.062396  4.979277  9.263001 12.634407 12.693897 11.105495
## KDM5D    11.060686 10.000682  3.697632 10.991681  9.804978 10.061677
## THBS1     8.743274 11.188168 10.928276  9.001494  9.818238 10.756625
```

Class information is not used, so the filtering is fair.

---

## ClassifyR

Introduce it.
    </textarea>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script src="styling/remark-zoom.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"ratio": "16:9",
"navigation": {
"scroll": false
}
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function() {
  var d = document, s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})();</script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
  }
});
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
